### PR TITLE
Add extremely experimental gRPC-WSGI tests

### PR DESCRIPTION
**Pull request by public**
This...works but my library is barely 24 hours old so pretty rough around the edges and changing a lot right now.

I had to change the improbable test runner here to make it play nice with my Docker on Windows setup. I haven't tested the other frontends yet.

### CODE REVIEW

File: frontend/improbable/src/index.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ var client = new EchoServiceClient(host, {
 function makeEchoRequest() {
   const req = new EchoRequest();
   req.setMessage("test");
-  client.echo(req, function(
+  client.echo(req, function (
```
Comment (public) [inline frontend/improbable/src/index.ts]: These are all VSCode trying to be helpful. I can clean them out if you don't want the churn.

File: frontend/improbable/src/index.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ import {
   ServerStreamingEchoResponse
 } from "../proto/echo_pb";
 
-const host = "http://localhost:8080";
+const host = "http://" + window.location.hostname + ":8080";
```
Comment (public) [inline frontend/improbable/src/index.ts]: On Docker for Windows (and I guess also Mac) the Docker host IP isn't localhost, fortunately it's easy to just pull it out of the URL here.

File: backend/grpcwsgi/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+FROM python:3.7
+
+WORKDIR /usr/src/app
+
+COPY . .
+
+RUN ls -l
+
+RUN pip install git+git://github.com/public/grpcWSGI.git
```
Comment (public) [inline backend/grpcwsgi/Dockerfile]: Library is extremely new so not on PyPI yet.

File: frontend/improbable/src/index.ts (hunk 2 above)
Comment (johanbrandhorst) [reply frontend/improbable/src/index.ts]: TIL, good fix!

**Comment by johanbrandhorst** [comment]
Ah, I checked out the branch and I'm a little confused. I see you put `grpcwsgi` in `./backend/`. Help me understand how `grpcwsgi` relates to a normal gRPC server.

* Is it only possible to use with python gRPC backends (like in the example here)?
* Is it possible to run standalone and just wrap a gRPC address (like the other two proxies, except the `inprocess`)

I think we should move `grpcwsgi` into the `proxy` folder anyway (the `inprocess` proxy is a precedent for this). What do you think?

**Comment by public** [comment]
So grpcwsgi has to work by replacing pretty much the entire grpc core IO implementation since it's all hidden inside compiled C extensions that have very limited extensibility from pure Python.

I understand they did it this way to reduce the number of ground up gRPC implementations they need to maintain but the way they've done it doesn't play that nicely with others.

This means it's not really a proxy since it's not transforming requests from the layer underneath, it has to replace everything to work at all. The only bit grpcWSGI keeps is the APIs that come out of the generated code which is mostly just declaring the services and their RPCs.

I'm happy to pretend it's a proxy if that makes sense to you though.

**Comment by johanbrandhorst** [comment]
AH, so it doesn't actually transform the grpc-web requests to gRPC and then pass it along to the gRPC-Python implementation? It just exposes a gRPC-Web API externally and a gRPC-python _like_ API internally?

Hm, I see the confusion. I think for the purposes of this repo it would still be considered a "proxy" though. I think we should add another little definition in the README explaining what we consider to be proxies; namely, things that handle gRPC-Web requests in some way (either proxying them downstream over the nextwork or by handling them internally).

Does that sound okay?

**Comment by public** [comment]
Yep that's exactly it. I might at some point dig through how the C bindings work these days and see if it's possible to use the C gRPC library underneath but not yet.

I'll move it to the proxy folder 👍

File: README.md [hunk 4]
```diff
@@ -56,6 +56,12 @@ Test various implementations of gRPC-Web Clients with various implementations of
   available at https://github.com/envoyproxy/envoy/tree/master/source/extensions/filters/http/grpc_web.
   It supports XHR and both the `application/grpc-web` and `application/grpc-web-text` content-types.
 
+- `grpcwsgi``
+
+    A Python WSGI compatible implementation of gRPC-Web, available at https://github.com/public/grpcWSGI.
+    It supports Fetch/XHR only.
+    It supports `application/grpc-web` only.
```
Comment (johanbrandhorst) [inline README.md]: Nit: the rest of this document uses 2 spaces for indentation

File: README.md [hunk 5]
```diff
@@ -92,6 +98,7 @@ it includes the server itself.
 | `envoy`        | ✔️           | ✔️️       | ✔️            | ❌                  |
 | `grpcwebproxy` | ✔️️          | ✔️        | ✔️            | ✔️️                 |
 | `inprocess` | ✔️️          | ✔️        | ✔️            | ✔️️                 |
+| `grpcwsgi` | ✔️️          | ✔️        | ✔️            | ❌                 |
```
Comment (johanbrandhorst) [inline README.md]: This should have an `X` in the `grpcwebtext` column.

File: README.md [hunk 6]
```diff
@@ -56,6 +56,12 @@ Test various implementations of gRPC-Web Clients with various implementations of
   available at https://github.com/envoyproxy/envoy/tree/master/source/extensions/filters/http/grpc_web.
   It supports XHR and both the `application/grpc-web` and `application/grpc-web-text` content-types.
 
+- `grpcwsgi``
```
Comment (johanbrandhorst) [inline README.md]: Extra tick?

File: README.md (hunk 5 above)
Comment (johanbrandhorst) [reply README.md]: Should add  `grpcwsgi` to line 91 as well.

Comment (johanbrandhorst) [reply README.md]: The last comment was missed in the latest update.

Comment (johanbrandhorst) [reply README.md]: Nevermind, there it is!

File: README.md [hunk 7]
```diff
@@ -82,7 +88,7 @@ Note: The websocket transport is not part of the grpc-web spec.
    $ docker-compose down
    ```
 
-Note: The `inprocess` proxy does not require `echo-server` to be running,
+Note: The `inprocess` and `grpcwsgi` proxies do not require `echo-server` to be running,
 it includes the server itself.
```
Comment (johanbrandhorst) [inline README.md]:
```suggestion
they include the server themselves.
```

**Pull request merged**